src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs

```diff
@@ -30,13 +31,19 @@ protected override void OnReadStarting()
             }
         }
 
-        protected override void OnReadStarted()
+        protected override Task OnReadStartedAsync()
         {
             // Produce 100-continue if no request body data for the stream has arrived yet.
             if (!_context.RequestBodyStarted)
             {
-                TryProduceContinue();
+                ValueTask<FlushResult> continueTask = TryProduceContinueAsync();
+                if (!continueTask.IsCompletedSuccessfully)
+                {
+                    return continueTask.AsTask();
+                }
```

Comment:
We have a helper method for `ValueTask<FlushResult>` in this situation. Something like `continueTask.GetAsTask()`

Reply:
https://github.com/dotnet/aspnetcore/blob/d94cee250d7df317fa5b87b1a2396ff9dbb08283/src/Shared/ValueTaskExtensions/ValueTaskExtensions.cs

Reply:
```suggestion
                return TryProduceContinueAsync().GetAsTask();
```